src/base/machine.F

```diff
@@ -33,6 +33,25 @@ MODULE machine
   INTEGER, PUBLIC, PARAMETER                   :: default_output_unit = output_unit, &
                                                   default_input_unit  = input_unit
 
+  ! Enumerates the target architecture or instruction set extension.
+  ! A feature is present if within range for the respective architecture.
+  ! For example, to check for MACHINE_X86_AVX the following is true:
+  ! MACHINE_X86_AVX <= m_cpuid() and MACHINE_X86 >= m_cpuid().
+  ! For example, to check for MACHINE_ARM_SOME the following is true:
+  ! MACHINE_ARM_SOME <= m_cpuid() and MACHINE_ARM >= m_cpuid().
+  INTEGER, PARAMETER :: &
+    MACHINE_CPU_GENERIC =    0, &
+    MACHINE_X86_SSE4    = 1000, &
+    MACHINE_X86_AVX     = 1001, &
+    MACHINE_X86_AVX2    = 1002, &
+    MACHINE_X86_AVX512  = 1003, &
+    MACHINE_X86 = MACHINE_X86_AVX512 ! marks end of range
+   ! other arch to be added as needed e.g.,
+   !MACHINE_ARM_SOME    = 2000
+   !MACHINE_ARM_ELSE    = 2001
+   !MACHINE_ARM = MACHINE_ARM_ELSE
+   !MACHINE_PWR_????    = 3000
+
```

Comment:
personally I would prefer binary over decimal ranges, allowing for easier subdivision of ranges:

```fortran
  INTEGER, PARAMETER :: &
    MACHINE_CPU_GENERIC =    0, &
    MACHINE_X86         = z'1000', &
    MACHINE_X86_SSE4    = z'1001', &
    MACHINE_X86_AVX     = z'1010', &
    MACHINE_X86_AVX2    = z'1011', &
    MACHINE_X86_AVX512  = z'1012'
```

then the following would hold:

```fortran
IAND(m_cpuid(), MACHINE_X86) ! .TRUE. for any X86 flag
IAND(m_cpuid(), MACHINE_AVX) ! .TRUE. for any AVX flag
```

... assuming that a machine implementing AVX512 also implements AVX2 and AVX2, not sure whether that assumption is sensible.

But I guess you'd prefer to re-use what you introduced in `libxsmm` already.

Reply:
Well, reusing the stuff from LIBXSMM is not the reason for the decimal ranges. The background is that INTEGER or INTEGER(8) only yield 32- or 64-bits respectively. If we go for binary flags we perhaps exhaust the number of bits. Remember, the bits are shared between Intel, ARM, Power and what not. Changing the design to not rely on an INTEGER on the other hand may reduce the usability of cpuid functions. Though, bits are inefficient in this design since a certain say feature on ARM may not appear at the same time as e.g. Intel AVX (nevertheless both "features" would need distinct bits/flags).